### PR TITLE
Feat/stash unstaged

### PR DESCRIPTION
**Pull request by Julienlin**
Screenshot from 2021-05-21 11-50-14

### CODE REVIEW

File: src/commands/run.ts [hunk 1]
```diff
@@ -82,6 +82,13 @@ export function addRun(program: commander.Command): void {
         });
       }
 
+      const stashMessage = 'stashing unstaged changes in order to run hooks properly';
+      const unstashMessage = 'un' + stashMessage;
```
Comment (LMaxence) [inline src/commands/run.ts:86]: Could you write the message explicitely please ?

File: src/commands/run.ts [hunk 2]
```diff
@@ -82,6 +82,13 @@ export function addRun(program: commander.Command): void {
...
+      if (type === 'pre-commit') {
+        console.log(chalk.bgCyan(stashMessage));
+        execSync(`git stash push --keep-index --include-untracked -m "MOOKME: ${stashMessage}"`);
+      }
```
Comment (LMaxence) [inline src/commands/run.ts:90]: let's have this as a warning, so in yellow (`chalk.yellow.bold`) + list the stashed files :)
Also a good thing would to only display this if the workspace is not fully staged

File: src/commands/run.ts [hunk 3]
```diff
@@ -104,6 +111,12 @@ export function addRun(program: commander.Command): void {
         console.error(err);
       }
 
+      if (type === 'pre-commit') {
```
Comment (LMaxence) [inline src/commands/run.ts:114]: Same : 'pre-commit' is defined in the `hook-types.ts` constant

File: src/commands/run.ts [hunk 4]
```diff
@@ -104,6 +111,12 @@ export function addRun(program: commander.Command): void {
...
+        console.log();
+        console.log(chalk.bgCyan(unstashMessage));
+        execSync('git stash pop');
```
Comment (LMaxence) [inline src/commands/run.ts:117]: Same here, yellow + bold, + would be nice to not always display this if not needed

File: src/commands/run.ts (hunk 2 above)
Comment (Julienlin) [reply src/commands/run.ts:95]: Screenshot from 2021-06-04 10-02-24

Comment (Julienlin) [reply src/commands/run.ts:100]: Due to some internal configuration of git stash it is not trivial to retrieve the list of untracked files.

I have found [here](https://gist.github.com/imghasemi/015088c966a1ba441b30881816e0d94e) a way to retrieve this list.

Here is now what the output looks like : 
Screenshot from 2021-06-04 11-30-16

File: src/commands/run.ts (hunk 4 above)
Comment (Julienlin) [reply src/commands/run.ts:127]: Resolved in 0e1647dca9a5748ac6f7ce9b540144e526e878f1

File: src/commands/run.ts (hunk 2 above)
Comment (LMaxence) [reply src/commands/run.ts:100]: And echo the stash command before we execute it : 
`console.log(chalk.bold(\`> git stash push --keep-index --include-untracked -m "MOOKME: ${stashMessage}"\`))`

Comment (Julienlin) [reply src/commands/run.ts:101]: Resolved in a2f1f5f

**Comment by Julienlin** [comment]
> What happens if there is no saved files here ? This should return an empty list, then throw at `split('\n').slice(2)` right ?

A funny thing happen, here is the output of the command I used to retrieve untracked stashed files : 
```
fatal: Invalid object name 'stash@{0}^3'.

```
There is always 2 `\n` in the ouput

**Pull request merged**